### PR TITLE
Make the doSave method compatible with Symfony\Component\Cache\Traits\AbstractTrait::doSave

### PR DESCRIPTION
**Pull request by emodric**
Current implementation is working fine, but in theory is not correct as it does not follow the rules of contravariance.

Moreover, it fails in PHP 8 (I am testing eZ Platform 2.5 on PHP 8 with `--ignore-platform-reqs`) as PHP 8 introduces the validation of abstract trait methods:

https://3v4l.org/hVGRg

Can this be merged and tagged since it does not introduce any breaking change and allows me to play with eZ Platform 2.5 on PHP 8?

Thanks!

### CODE REVIEW

**Comment by emodric** [comment]
Ping @andrerom @alongosz

**Review by alongosz — CHANGES_REQUESTED** [review]
Tests are failing here.
It doesn't seem like it's compatible with Symfony 3.4 for which this package was created AFAIR, see https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/Cache/Traits/AbstractTrait.php#L80

**Comment by emodric** [comment]
Tests themseves are passing:

image

The issue with the build seems like a problem with PHP CS Fixer.

The method signature in Symfony 3.4 is `abstract protected function doSave(array $values, $lifetime);` which IS compatible with the new signature from this PR: `abstract protected function doSave(array $values, $lifetime, array $addTagData = [], array $removeTagData = []): array;`. Added arguments do not cause issues.

**Comment by Steveb-p** [comment]
@alongosz It's from what I see a backport from Symfony 4.x, with parameter `$lifetime` being narrowed down to `int`. Type narrowing from undeclared type is allowed for PHP7, but prohibited by PHP8.

It looks good to me, I just didn't have the time to read into this one properly.

To be frank since this is a polyfill class it should be wrapped with `class_exists` check to prevent collision with Symfony classes it intends to polyfill, just in case, but that's beside the point.

**Comment by Steveb-p** [comment]
@emodric to be on the safe side, could you please provide me with a detailed (relatively) error with stacktrace, so I can reproduce it in tests, and then attach them here?

I'm already setting up tests for PHP8 in some other packages of ours.

In the meantime until this is merged, could you please use your fork (with normal composer's repository overwrite, as per their docs)?

**Comment by emodric** [comment]
@Steveb-p Actually, there's no stack trace since this is a fatal compile error, but the error message is:

```
[Symfony\Component\Debug\Exception\FatalErrorException]                                                                                                                                                         
Compile Error: Declaration of Symfony\Component\Cache\Adapter\AbstractTagAwareAdapter::doSave(array $values,
int $lifetime, array $addTagData = [], array $removeTagData = []): array must be compatible with
Symfony\Component\Cache\Traits\AbstractTrait::doSave(array $values, $lifetime)                                                                                                                                                                                                                                                                                                                                                    

at /var/www/html/ezplatform2/vendor/ezsystems/symfony-tools/
src/lib/Symfony/Components/Cache/Adapter/AbstractTagAwareAdapter.php:148
```

**Comment by Steveb-p** [comment]
Added a branch with tests that show this particular issue, along with new CI pipeline: #13.

**Comment by Steveb-p** [comment]
@emodric please rebase the branch on the current `master` and tests should lit up green.

**Comment by emodric** [comment]
@Steveb-p Done!

**Comment by Steveb-p** [comment]
Seems like some test classes also have invalid `setUp` declarations, which would blow the tests up. Can you fix those? If not I'll do that myself, in the evening.

**Comment by emodric** [comment]
It looks like the tests that blow up actually come from Symfony.

**Comment by Steveb-p** [comment]
We will need to use https://symfony.com/doc/current/components/phpunit_bridge.html#removing-the-void-return-type then.

Probably introducing trait usage here would be the least amount of work?

**Comment by emodric** [comment]
Hm... Trait requires us to redefine `setUp` and friends as `doSetUp`, and since original methods are in `vendor/`, I don't think trait would help us.

I've added `SYMFONY_PHPUNIT_REMOVE_RETURN_TYPEHINT` environment variable instead.

Let's see what happens now.

**Comment by Steveb-p** [comment]
I'm pretty sure PHPStan will complain about those methods. It's not aware that PHPUnit Symfony Bridge does anything with method declarations.

**Comment by emodric** [comment]
I can try working around it somehow, but do we really need PHPStan here, since the repo is pretty much is maintenance mode?

**Comment by Steveb-p** [comment]
@emodric it makes sense to have it, even at this stage. There are already inconsistencies detected for ezplatform-kernel that were found by it.

Let me see what I can do in the evening. Worst case I'll disable PHPStan checks on PHP8 and we'll see how it goes.

**Comment by emodric** [comment]
Hi @Steveb-p . Any progress on this?

Thanks!

**Comment by Steveb-p** [comment]
> Hi @Steveb-p . Any progress on this?
> 
> Thanks!

I forgot to get back to this. I'll set a reminder for myself and will take a look tomorrow evening.

**Comment by Steveb-p** [comment]
@emodric sorry for the late response. I've moved your email notification (GH) into important directory in my mailbox, and then wasn't able to find it, because I've forgotten I've moved it.

Derp.

The solution for the CI issue is simple: remove 9.x PHPUnit from the list of `require-dev`. This will cause the version 7.x to be installed, which is compatible in every way with symfony/cache 3.x.

This should make the tests green. I've tested it against PHP 7.4.

**Comment by Steveb-p** [comment]
Travis had some sort of a hiccup and needed the job to be restarted. Tests are green now.

**Review by adamwojs — APPROVED** [review]

**Review by alongosz — APPROVED** [review]

File: phpunit.xml.dist [hunk 1]
```diff
@@ -12,6 +12,7 @@
         <ini name="error_reporting" value="-1" />
         <env name="REDIS_HOST" value="localhost" />
         <env name="MEMCACHED_HOST" value="localhost" />
+        <env name="SYMFONY_PHPUNIT_REMOVE_RETURN_TYPEHINT" value="1" />
```
Comment (Steveb-p) [inline phpunit.xml.dist:15]: No longer needed.

```suggestion
```

**Comment by emodric** [comment]
@Steveb-p Thanks for seeing this to the end :) I suppose you will tag 1.0.5 as soon as this is merged?

**Comment by Steveb-p** [comment]
@emodric Probably something like that - it's technically a bugfix. There is hardly a BC break, since it wasn't working on PHP8 anyway (where covariance and contravariance requirement is hardened).

I'm not usually the one doing the version releases though.

**Pull request merged**